File: src/renderer/particles.c
#include "../types.h"
#include "gl_backend.h"
#include "transform.h"
#include <glad/glad.h>
#include <string.h>

typedef struct {
    float x, y, z;
    float vx, vy, vz;
    float life;
    float max_life;
    uint8 type;
    bool active;
} Particle;

#define MAX_PARTICLES 256
static Particle s_particles[MAX_PARTICLES];

/* GL resources for a small quad used to draw each particle */
static GLuint s_quad_vao;
static GLuint s_quad_vbo;

/* Build a translation + uniform-scale model matrix (column-major 4x4) */
static void BuildParticleModel(float *m, float x, float y, float z, float scale) {
    /*  scale  0      0      0
        0      scale  0      0
        0      0      scale  0
        x      y      z      1   */
    memset(m, 0, 16 * sizeof(float));
    m[0]  = scale;
    m[5]  = scale;
    m[10] = scale;
    m[12] = x;
    m[13] = y;
    m[14] = z;
    m[15] = 1.0f;
}

void Particles_Init(void) {
    memset(s_particles, 0, sizeof(s_particles));

    /* Unit quad centred at origin, in XY plane.
       Vertices ordered for GL_TRIANGLE_FAN. */
    static const float quad_verts[] = {
        -0.5f, -0.5f, 0.0f,
         0.5f, -0.5f, 0.0f,
         0.5f,  0.5f, 0.0f,
        -0.5f,  0.5f, 0.0f,
    };

    glGenVertexArrays(1, &s_quad_vao);
    glGenBuffers(1, &s_quad_vbo);

    glBindVertexArray(s_quad_vao);
    glBindBuffer(GL_ARRAY_BUFFER, s_quad_vbo);
    glBufferData(GL_ARRAY_BUFFER, sizeof(quad_verts), quad_verts, GL_STATIC_DRAW);

    /* aPos — location 0 in g_flat_shader */
    glEnableVertexAttribArray(0);
    glVertexAttribPointer(0, 3, GL_FLOAT, GL_FALSE, 3 * sizeof(float), (void *)0);

    glBindVertexArray(0);
}

void Particles_Shutdown(void) {
    memset(s_particles, 0, sizeof(s_particles));

    if (s_quad_vbo) { glDeleteBuffers(1, &s_quad_vbo); s_quad_vbo = 0; }
    if (s_quad_vao) { glDeleteVertexArrays(1, &s_quad_vao); s_quad_vao = 0; }
}

void Particles_Spawn(float x, float y, float z,
                     float vx, float vy, float vz,
                     float life, uint8 type) {
    for (int i = 0; i < MAX_PARTICLES; i++) {
        if (!s_particles[i].active) {
            s_particles[i] = (Particle){
                .x = x, .y = y, .z = z,
                .vx = vx, .vy = vy, .vz = vz,
                .life = life, .max_life = life,
                .type = type, .active = true
            };
            return;
        }
    }
}

void Particles_Update(float dt) {
    if (dt <= 0.0f) return;
    for (int i = 0; i < MAX_PARTICLES; i++) {
        Particle *p = &s_particles[i];
        if (!p->active) continue;
        p->life -= dt;
        if (p->life <= 0.0f) {
            p->active = false;
            continue;
        }
        p->x += p->vx * dt;
        p->y += p->vy * dt;
        p->z += p->vz * dt;
    }
}

void Particles_Render(void) {
    if (!g_flat_shader) return;

    /* Count active particles first to avoid GPU state changes when idle */
    int any_active = 0;
    for (int i = 0; i < MAX_PARTICLES; i++) {
        if (s_particles[i].active) { any_active = 1; break; }
    }
    if (!any_active) return;

    const float *view = Transform_GetView();
    const float *proj = Transform_GetProjection();

    glUseProgram(g_flat_shader);
    GlBackend_SetMat4(g_flat_shader, "uView", view);
    GlBackend_SetMat4(g_flat_shader, "uProj", proj);

    /* Particles are additive-blended, depth-tested but not depth-written */
    glEnable(GL_BLEND);
    glBlendFunc(GL_SRC_ALPHA, GL_ONE);
    glDepthMask(GL_FALSE);

    glBindVertexArray(s_quad_vao);

    for (int i = 0; i < MAX_PARTICLES; i++) {
        const Particle *p = &s_particles[i];
        if (!p->active) continue;

        float alpha = (p->max_life > 0.0f) ? (p->life / p->max_life) : 0.0f;
        if (alpha <= 0.0f) continue;

        /* Colour by type */
        float r, g, b;
        switch (p->type) {
        default: /* fall through */
        case 0: r = 1.0f; g = 0.4f; b = 0.1f; break;  /* fire (orange/red) */
        case 1: r = 1.0f; g = 1.0f; b = 0.2f; break;  /* spark (yellow)    */
        case 2: r = 1.0f; g = 1.0f; b = 1.0f; break;  /* flash (white)     */
        case 3: r = 0.3f; g = 0.5f; b = 1.0f; break;  /* energy (blue)     */
        }

        float model[16];
        BuildParticleModel(model, p->x, p->y, p->z, 2.0f);

        GlBackend_SetMat4(g_flat_shader, "uModel", model);
        GlBackend_SetVec4(g_flat_shader, "uColor", r, g, b, alpha);

        glDrawArrays(GL_TRIANGLE_FAN, 0, 4);
    }

    glBindVertexArray(0);

    /* Restore state */
    glDepthMask(GL_TRUE);
    glDisable(GL_BLEND);
}

File: rust/sf-render/src/particles.rs
//! Additive-blended billboard particles.
//!
//! Port (C oracle): `src/renderer/particles.c`.

use crate::gpu::{Gpu, Vertex3};
use crate::transform::Transform;

pub const MAX_PARTICLES: usize = 256;

#[derive(Debug, Clone, Copy, Default)]
struct Particle {
    x: f32,
    y: f32,
    z: f32,
    vx: f32,
    vy: f32,
    vz: f32,
    life: f32,
    max_life: f32,
    ptype: u8,
    active: bool,
}

/// Build a translation + uniform-scale model matrix (column-major 4x4).
fn build_particle_model(x: f32, y: f32, z: f32, scale: f32) -> [f32; 16] {
    let mut m = [0.0f32; 16];
    m[0] = scale;
    m[5] = scale;
    m[10] = scale;
    m[12] = x;
    m[13] = y;
    m[14] = z;
    m[15] = 1.0;
    m
}

pub struct Particles {
    particles: [Particle; MAX_PARTICLES],
    /// Unit quad centred at origin (XY plane), the GL_TRIANGLE_FAN expanded
    /// into a 6-vertex triangle list for the retained flat pipeline.
    quad_tris: [Vertex3; 6],
}

impl Particles {
    pub fn new(_gpu: &mut Gpu) -> Self {
        // Unit quad centred at origin, in XY plane. Fan order was
        // (v0,v1,v2,v3); triangulate as (v0,v1,v2) + (v0,v2,v3).
        let v0 = Vertex3 {
            pos: [-0.5, -0.5, 0.0],
        };
        let v1 = Vertex3 {
            pos: [0.5, -0.5, 0.0],
        };
        let v2 = Vertex3 {
            pos: [0.5, 0.5, 0.0],
        };
        let v3 = Vertex3 {
            pos: [-0.5, 0.5, 0.0],
        };
        Particles {
            particles: [Particle::default(); MAX_PARTICLES],
            quad_tris: [v0, v1, v2, v0, v2, v3],
        }
    }

    #[allow(clippy::too_many_arguments)]
    pub fn spawn(
        &mut self,
        x: f32,
        y: f32,
        z: f32,
        vx: f32,
        vy: f32,
        vz: f32,
        life: f32,
        ptype: u8,
    ) {
        for p in self.particles.iter_mut() {
            if !p.active {
                *p = Particle {
                    x,
                    y,
                    z,
                    vx,
                    vy,
                    vz,
                    life,
                    max_life: life,
                    ptype,
                    active: true,
                };
                return;
            }
        }
    }

    pub fn update(&mut self, dt: f32) {
        if dt <= 0.0 {
            return;
        }
        for p in self.particles.iter_mut() {
            if !p.active {
                continue;
            }
            p.life -= dt;
            if p.life <= 0.0 {
                p.active = false;
                continue;
            }
            p.x += p.vx * dt;
            p.y += p.vy * dt;
            p.z += p.vz * dt;
        }
    }

    /// Mirror of `Particles_Render`. NOTE: the old pass was additive-blended
    /// and depth-write-disabled; the retained flat pipeline offers neither, so
    /// particles now draw as opaque depth-writing quads (see parity note).
    pub fn render(&self, gpu: &mut Gpu, transform: &Transform) {
        if !self.particles.iter().any(|p| p.active) {
            return;
        }

        let proj = *transform.projection();
        let view = *transform.view();

        for p in &self.particles {
            if !p.active {
                continue;
            }
            let alpha = if p.max_life > 0.0 {
                p.life / p.max_life
            } else {
                0.0
            };
            if alpha <= 0.0 {
                continue;
            }

            // Colour by type
            let (r, g, b) = match p.ptype {
                1 => (1.0, 1.0, 0.2), // spark (yellow)
                2 => (1.0, 1.0, 1.0), // flash (white)
                3 => (0.3, 0.5, 1.0), // energy (blue)
                _ => (1.0, 0.4, 0.1), // fire (orange/red)
            };

            let model = build_particle_model(p.x, p.y, p.z, 2.0);
            // Additive, non-depth-writing (was GL SRC_ALPHA/ONE + depth mask
            // off) so particles glow and layer instead of occluding.
            gpu.push_flat_tris_add(&self.quad_tris, &proj, &view, &model, [r, g, b, alpha]);
        }
    }
}
